tachyon/zk/air/expressions/evaluator.h

```diff
@@ -0,0 +1,112 @@
+#ifndef TACHYON_ZK_AIR_EXPRESSIONS_EVALUATOR_H_
+#define TACHYON_ZK_AIR_EXPRESSIONS_EVALUATOR_H_
+
+#include <optional>
+#include <vector>
+
+#include "third_party/eigen3/Eigen/Core"
+
+#include "tachyon/base/logging.h"
+#include "tachyon/math/matrix/matrix_types.h"
+#include "tachyon/zk/air/expressions/first_row_expression.h"
+#include "tachyon/zk/air/expressions/last_row_expression.h"
+#include "tachyon/zk/air/expressions/transition_expression.h"
+#include "tachyon/zk/air/expressions/variable_expression.h"
+#include "tachyon/zk/expressions/expression.h"
+
+namespace tachyon::zk::air {
+
+template <typename F>
+class Evaluator {
+ public:
+  F Evaluate(const Expression<F>* input, Eigen::Index current_row,
+             Eigen::Index num_rows, const std::vector<F>& public_values,
+             std::optional<const Eigen::Ref<const math::RowMajorMatrix<F>>>
+                 preprocessed_window,
+             const math::RowMajorMatrix<F>& main_window) {
+    switch (input->type()) {
+      case ExpressionType::kConstant:
+        return input->ToConstant()->value();
+      case ExpressionType::kSum: {
+        const SumExpression<F>* sum = input->ToSum();
+        return Evaluate(sum->left(), current_row, num_rows, public_values,
+                        preprocessed_window, main_window) +
+               Evaluate(sum->right(), current_row, num_rows, public_values,
+                        preprocessed_window, main_window);
+      }
+      case ExpressionType::kProduct: {
+        const ProductExpression<F>* product = input->ToProduct();
+        return Evaluate(product->left(), current_row, num_rows, public_values,
+                        preprocessed_window, main_window) *
+               Evaluate(product->right(), current_row, num_rows, public_values,
+                        preprocessed_window, main_window);
+      }
+      case ExpressionType::kNegated: {
+        const NegatedExpression<F>* negated = input->ToNegated();
+        return -Evaluate(negated->expr(), current_row, num_rows, public_values,
+                         preprocessed_window, main_window);
+      }
+      case ExpressionType::kScaled: {
+        const ScaledExpression<F>* scaled = input->ToScaled();
+        return Evaluate(scaled->expr(), current_row, num_rows, public_values,
+                        preprocessed_window, main_window) *
+               scaled->scale();
+      }
+      case ExpressionType::kFirstRow: {
+        if (current_row == 0) {
+          const FirstRowExpression<F>* first_row = input->ToFirstRow();
+          return Evaluate(first_row->expr(), current_row, num_rows,
+                          public_values, preprocessed_window, main_window);
+        }
+        return F::Zero();
+      }
+      case ExpressionType::kLastRow: {
+        if (current_row == num_rows - 2) {
```

Comment:
In `ConstraintSystem::IsSatisfied()`, it only iterates until `num_rows - 2` while in Plonky3, it iterates until `num_rows - 1` but [the second row of the window points to first row of the trace](https://github.com/Plonky3/Plonky3/blob/7c5deb0eab7191a97f7bb088637c1a68b2e6eb68/uni-stark/src/check_constraints.rs#L19) in the final iteration. I found it kind of redundant since we can just use the last transition window $(row_{\text{numrows}-2}, row_{numrows-1})$ for enforcing `LastRowConstraint`.